tls/s2n_x509_validator.c

```diff
@@ -573,48 +574,32 @@ int s2n_disable_time_validation_ossl_verify_callback(int default_ossl_ret, X509_
     switch (err) {
         case X509_V_ERR_CERT_NOT_YET_VALID:
         case X509_V_ERR_CERT_HAS_EXPIRED:
-            return OSSL_VERIFY_CALLBACK_IGNORE_ERROR;
+            return OSSL_VERIFY_CALLBACK_OK;
         default:
             break;
     }
 
-    /* If CRL validation is enabled, setting the time validation verify callback will override the
-     * CRL verify callback. The CRL verify callback is manually triggered to work around this
-     * issue.
-     *
-     * The CRL verify callback ignores validation errors exclusively for CRL timestamp fields. So,
-     * if CRL validation isn't enabled, the CRL verify callback is a no-op.
-     */
-    return s2n_crl_ossl_verify_callback(default_ossl_ret, ctx);
+    return default_ossl_ret;
 }
 
-static S2N_RESULT s2n_x509_validator_disable_time_validation(struct s2n_connection *conn,
-        struct s2n_x509_validator *validator)
+int s2n_ossl_verify_callback(int default_ossl_ret, X509_STORE_CTX *ctx)
 {
-    RESULT_ENSURE_REF(conn);
-    RESULT_ENSURE_REF(conn->config);
-    RESULT_ENSURE_REF(validator);
-    RESULT_ENSURE_REF(validator->store_ctx);
-
-    /* Setting an X509_STORE verify callback is not recommended with AWS-LC:
-     * https://github.com/aws/aws-lc/blob/aa90e509f2e940916fbe9fdd469a4c90c51824f6/include/openssl/x509.h#L2980-L2990
-     *
-     * If the libcrypto supports the ability to disable time validation with an X509_VERIFY_PARAM
-     * NO_CHECK_TIME flag, this method is preferred.
-     *
-     * However, older versions of AWS-LC and OpenSSL 1.0.2 do not support this flag. In this case,
-     * an X509_STORE verify callback is used. This is acceptable in older versions of AWS-LC
-     * because the versions are fixed, and updates to AWS-LC will not break the callback
-     * implementation.
-     */
-    if (s2n_libcrypto_supports_flag_no_check_time()) {
-        RESULT_GUARD(s2n_x509_validator_set_no_check_time_flag(validator));
-    } else {
-        X509_STORE_CTX_set_verify_cb(validator->store_ctx,
-                s2n_disable_time_validation_ossl_verify_callback);
+    void *conn_parameter = X509_STORE_CTX_get_app_data(ctx);
+    if (conn_parameter == NULL) {
+        /* error, fail closed */
+        return OSSL_VERIFY_CALLBACK_ERROR;
     }
+    const struct s2n_connection *conn = (struct s2n_connection *) conn_parameter;
 
-    return S2N_RESULT_OK;
+    /* If any callback returns OK, then return OK */
+    if ((conn->config->disable_x509_time_validation && !s2n_libcrypto_supports_flag_no_check_time())
+            && s2n_disable_time_validation_ossl_verify_callback(default_ossl_ret, ctx) == OSSL_VERIFY_CALLBACK_OK) {
+        return OSSL_VERIFY_CALLBACK_OK;
+    }
```

Comment:
These aren't really being used as openssl verify callbacks anymore, they're more just being used to determine if we need to skip the error. I wonder if it would help to indicate this with the function name/signature, rather than keeping the openssl one? Not sure about the name, but maybe something like this?
```
bool s2n_x509_validator_should_skip_ossl_time_verify_error(X095_STORE_CTX *ctx);
```

I also wonder if it could help to have these functions take the s2n_connection and do the config/libcrypto checks necessary to determine if the error should be skipped?

Reply:
Adopted the new naming!

I don't know that the checks have enough complexity to make sense in a different method. Maybe take a look at the new functions and let me know if you still want me change it?

Reply:
No yeah the suggestion was just to put the checks in the new is_error functions. How you have it makes sense too though.